cmd/generic_chain.go

```diff
@@ -97,21 +104,93 @@ func pruneStateStore(stateStoreDB db.DB, pruneHeight uint64) error {
 	)
 }
 
-func pruneAppStore(appStore db.DB, pruneHeight uint64) error {
-	return pruneKeys(
-		appStore, "application", appKeyInfos,
-		func(store db.DB, key string) (uint64, error) {
-			return deleteHeightRange(store, key, 0, pruneHeight-1, asciiHeightParser)
-		},
-	)
+func pruneValidatorHistory(stateStoreDB db.DB, validatorHistoryToKeep uint64) error {
+	latestValHeight, err := findLatestValidatorHeight(stateStoreDB)
+	if err != nil {
+		return fmt.Errorf("could not determine latest validator height: %w", err)
+	}
+
+	if latestValHeight <= validatorHistoryToKeep {
+		logger.Info("not enough history to perform validatorKey pruning")
+		return nil
+	}
+
+	retainHeight := latestValHeight - validatorHistoryToKeep
+	anchorHeight, err := findAnchorCheckpoint(stateStoreDB, retainHeight)
+	if err != nil {
+		return fmt.Errorf("could not find safe anchor checkpoint: %w", err)
+	}
+
+	validatorPruneCutoff := anchorHeight - 1
+	if validatorPruneCutoff == 0 {
+		logger.Info("no old validator keys to prune")
+		return nil
+	}
+
+	logger.Info("pruning validatorsKey", "cutoffHeight", validatorPruneCutoff)
+	count, err := deleteHeightRange(stateStoreDB, "validatorsKey:", 0, validatorPruneCutoff, asciiHeightParser)
+	if err != nil {
+		return fmt.Errorf("failed to prune validatorsKey: %w", err)
+	}
+
+	logger.Info("pruned validatorsKey", "count", count)
+	return nil
 }
 
-func pruneBlockAndStateStore(blockStoreDB, stateStoreDB, appStore db.DB, pruneHeight uint64) error {
+func findLatestValidatorHeight(db db.DB) (uint64, error) {
+	prefix := []byte("validatorsKey:")
+	endPrefix := make([]byte, len(prefix))
+	copy(endPrefix, prefix)
+	endPrefix[len(endPrefix)-1]++
+
+	iter, err := db.ReverseIterator(prefix, endPrefix)
+	if err != nil {
+		return 0, err
+	}
+	defer func() {
+		_ = iter.Close()
+	}()
+
+	if !iter.Valid() {
+		if err := iter.Error(); err != nil {
+			return 0, err
+		}
+		return 0, fmt.Errorf("no validator keys found")
+	}
+
+	key := iter.Key()
+	keyStr := string(key)
+	heightStr := strings.TrimPrefix(keyStr, "validatorsKey:")
+	return strconv.ParseUint(heightStr, 10, 64)
+}
+
+// findAnchorCheckpoint searches backwards from retainHeight to find a height that has complete validator set data
+// in validatorsKey there are two types of values: a validator set, and a value saying that the complete validator set for this height is unchanged, and points to some other key.
+func findAnchorCheckpoint(db db.DB, retainHeight uint64) (uint64, error) {
```

Comment:
this is nice, i guess missing this is why celestia was panicking, we should try it again.

we should also be able to say `validatorHistoryToKeep = 0` right? as long as we find an anchor